File: backend/template_engine.py
"""
Template engine for dummy EPG title/description rendering.

Supports:
  - Placeholders          {name}
  - Chained pipes         {name|uppercase|trim|strip:-}
  - Conditionals          {if:group}body{/if}
                          {if:group=value}body{/if}
                          {if:group~regex}body{/if}
  - Legacy suffix         {name_normalize}  (lowercase, strip non-alphanumeric)

Safety:
  Template length, individual group values, and user-supplied regex patterns
  are all length-capped so a malicious config can't trigger catastrophic
  backtracking against channel-name inputs. Invalid regex patterns inside a
  conditional evaluate to false rather than raising.
"""
from __future__ import annotations

import re
from typing import Any, Optional

import safe_regex


class TemplateSyntaxError(ValueError):
    """Raised when a template is malformed or references an unknown transform."""


# ---------------------------------------------------------------------------
# Legacy helper (preserved for {name_normalize} back-compat with the previous
# applyTemplate() in DummyEPGSourceModal.tsx).
# ---------------------------------------------------------------------------
_NORMALIZE_STRIP_RE = re.compile(r"[^a-z0-9]")


def _legacy_normalize(value: str) -> str:
    return _NORMALIZE_STRIP_RE.sub("", value.lower())


# ---------------------------------------------------------------------------
# Transforms. Each takes (value, arg) — arg is None for arg-less transforms.
# ---------------------------------------------------------------------------
def _t_uppercase(value: str, _arg: Optional[str]) -> str:
    return value.upper()


def _t_lowercase(value: str, _arg: Optional[str]) -> str:
    return value.lower()


def _t_titlecase(value: str, _arg: Optional[str]) -> str:
    return value.title()


def _t_trim(value: str, _arg: Optional[str]) -> str:
    return value.strip()


def _t_strip(value: str, arg: Optional[str]) -> str:
    return value.strip(arg) if arg else value.strip()


def _t_replace(value: str, arg: Optional[str]) -> str:
    # `replace:<from>:<to>` — arg is "from:to" (first colon splits).
    if arg is None:
        return value
    if ":" in arg:
        src, dst = arg.split(":", 1)
    else:
        src, dst = arg, ""
    return value.replace(src, dst)


def _t_normalize(value: str, _arg: Optional[str]) -> str:
    return _legacy_normalize(value)


_TRANSFORMS = {
    "uppercase": _t_uppercase,
    "lowercase": _t_lowercase,
    "titlecase": _t_titlecase,
    "trim": _t_trim,
    "strip": _t_strip,
    "replace": _t_replace,
    "normalize": _t_normalize,
}


class TemplateEngine:
    """Render dummy-EPG templates against extracted regex groups.

    Instances are stateless and cheap to construct, so they are freely
    reusable across renders.
    """

    # Public caps — exposed so tests and callers can reason about them.
    # The regex-length cap is inherited from ``safe_regex.DEFAULT_MAX_PATTERN_LEN``
    # (500 chars). Conditional regex patterns are routed through ``safe_regex``
    # which handles both the length cap and the ReDoS timeout.
    MAX_TEMPLATE_LEN = 4096
    MAX_INPUT_LEN = 1024

    # Pattern for a single {placeholder} — body has no unescaped '{' or '}'.
    _PLACEHOLDER_RE = re.compile(r"\{([^{}]+)\}")

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------
    def render(
        self,
        template: str,
        groups: dict[str, Any],
    ) -> str:
        if template is None:
            return ""
        if len(template) > self.MAX_TEMPLATE_LEN:
            raise TemplateSyntaxError(
                f"Template exceeds maximum length of {self.MAX_TEMPLATE_LEN} chars"
            )

        # Truncate group values up-front so every downstream transform and
        # regex conditional operates on bounded input.
        bounded_groups = {k: self._truncate(str(v)) for k, v in groups.items()}
        return self._render_segment(template, bounded_groups)

    def render_with_trace(
        self,
        template: str,
        groups: dict[str, Any],
    ) -> tuple[str, list[dict]]:
        """Render template and return (output, trace).

        Trace entries:
          - {"kind": "literal", "text": <str>}
          - {"kind": "placeholder", "raw": "{name|pipe}", "group_name": <str>,
              "initial_value": <str>, "pipes": [<pipe_step>...], "final_value": <str>}
          - {"kind": "conditional", "condition": <str>, "kind_detail": "truthy"|
              "equality"|"regex", "taken": <bool>, "value": <str>, "body": [...]}
        Pipe step: {"transform": <str>, "arg": <str|None>, "input": <str>,
                     "output": <str>, "source"?: <str> (provenance note, e.g.
                     "legacy _normalize suffix")}
        """
        if template is None:
            return "", []
        if len(template) > self.MAX_TEMPLATE_LEN:
            raise TemplateSyntaxError(
                f"Template exceeds maximum length of {self.MAX_TEMPLATE_LEN} chars"
            )
        bounded_groups = {k: self._truncate(str(v)) for k, v in groups.items()}
        trace: list[dict] = []
        output = self._render_segment(template, bounded_groups, trace_out=trace)
        return output, trace

    # ------------------------------------------------------------------
    # Core rendering
    # ------------------------------------------------------------------
    def _render_segment(
        self,
        template: str,
        groups: dict[str, str],
        trace_out: Optional[list[dict]] = None,
    ) -> str:
        out: list[str] = []
        i = 0
        n = len(template)

        while i < n:
            brace = template.find("{", i)
            if brace == -1:
                tail = template[i:]
                out.append(tail)
                if trace_out is not None and tail:
                    trace_out.append({"kind": "literal", "text": tail})
                break

            # Emit any leading literal text.
            if brace > i:
                literal = template[i:brace]
                out.append(literal)
                if trace_out is not None:
                    trace_out.append({"kind": "literal", "text": literal})

            close = template.find("}", brace)
            if close == -1:
                raise TemplateSyntaxError(f"Unclosed '{{' at position {brace}")

            directive = template[brace + 1 : close]

            if directive.startswith("if:"):
                body_start = close + 1
                body_end, after = self._find_matching_endif(template, body_start)
                condition = directive[3:]
                body_trace: list[dict] = []
                taken, detail = self._evaluate_condition(condition, groups)
                if taken:
                    out.append(self._render_segment(
                        template[body_start:body_end], groups,
                        trace_out=body_trace if trace_out is not None else None,
                    ))
                if trace_out is not None:
                    trace_out.append({
                        "kind": "conditional",
                        "condition": condition,
                        "kind_detail": detail["kind"],
                        "taken": taken,
                        "value": detail.get("value", ""),
                        "body": body_trace,
                    })
                i = after
                continue

            if directive == "/if":
                raise TemplateSyntaxError("Unmatched '{/if}'")

            out.append(self._render_placeholder(directive, groups, trace_out=trace_out))
            i = close + 1

        return "".join(out)

    # ------------------------------------------------------------------
    # {if:...} helpers
    # ------------------------------------------------------------------
    def _find_matching_endif(self, template: str, body_start: int) -> tuple[int, int]:
        """Return (body_end, after_endif) for the nearest matching {/if}.

        Handles nested conditionals by tracking {if:...} depth.
        """
        depth = 1
        i = body_start
        n = len(template)
        while i < n:
            b = template.find("{", i)
            if b == -1:
                break
            c = template.find("}", b)
            if c == -1:
                break
            directive = template[b + 1 : c]
            if directive.startswith("if:"):
                depth += 1
            elif directive == "/if":
                depth -= 1
                if depth == 0:
                    return b, c + 1
            i = c + 1
        raise TemplateSyntaxError("'{if:...}' without matching '{/if}'")

    def _evaluate_condition(
        self, condition: str, groups: dict[str, str]
    ) -> tuple[bool, dict]:
        """Evaluate a conditional. Returns (taken, detail) where detail describes
        the evaluation kind and the value examined, for trace display."""
        # Equality first so a '=' inside the regex doesn't get mis-parsed.
        if "=" in condition and "~" not in condition.split("=", 1)[0]:
            name, expected = condition.split("=", 1)
            value = groups.get(name, "")
            return value == expected, {"kind": "equality", "value": value}

        if "~" in condition:
            name, pattern = condition.split("~", 1)
            value = groups.get(name, "")
            # Fallback contract (bd-eio04.16): route user-supplied regex through
            # safe_regex. ``compile`` raises ``PatternTooLongError`` for oversize
            # patterns and ``SafeRegexError`` for malformed patterns — both are
            # caught below and collapse the conditional to false rather than
            # crashing the render. ``search`` on the compiled pattern returns
            # None on ReDoS timeout (logged by safe_regex), again collapsing to
            # false. A single bad user config therefore skips a {if:...~...}
            # block rather than bringing down XMLTV generation.
            try:
                compiled = safe_regex.compile(pattern)
            except safe_regex.PatternTooLongError:
                return False, {"kind": "regex", "value": value, "error": "pattern too long"}
            except safe_regex.SafeRegexError as err:
                return False, {"kind": "regex", "value": value, "error": str(err)}
            matched = safe_regex.search(compiled, value) is not None
            return matched, {"kind": "regex", "value": value}

        # {if:group} — truthy if non-empty.
        value = groups.get(condition, "")
        return bool(value), {"kind": "truthy", "value": value}

    # ------------------------------------------------------------------
    # {placeholder|pipe|...} rendering
    # ------------------------------------------------------------------
    def _render_placeholder(
        self,
        body: str,
        groups: dict[str, str],
        trace_out: Optional[list[dict]] = None,
    ) -> str:
        parts = body.split("|")
        name = parts[0].strip()
        pipes = parts[1:]
        raw = "{" + body + "}"

        # Legacy suffix: {name_normalize} → render with the legacy-normalize pipe.
        if name.endswith("_normalize") and name not in groups:
            base = name[: -len("_normalize")]
            initial = groups.get(base, "")
            final = _legacy_normalize(initial)
            if trace_out is not None:
                trace_out.append({
                    "kind": "placeholder",
                    "raw": raw,
                    "group_name": base,
                    "initial_value": initial,
                    "pipes": [{
                        "transform": "normalize",
                        "arg": None,
                        "input": initial,
                        "output": final,
                        "source": "legacy _normalize suffix",
                    }],
                    "final_value": final,
                })
            return final

        initial = groups.get(name, "")
        value = initial
        pipe_steps: list[dict] = []

        for pipe_spec in pipes:
            pipe_spec = pipe_spec.strip()
            if not pipe_spec:
                continue

            if ":" in pipe_spec:
                transform, arg = pipe_spec.split(":", 1)
            else:
                transform, arg = pipe_spec, None

            step_input = value
            fn = _TRANSFORMS.get(transform)
            if fn is None:
                raise TemplateSyntaxError(f"Unknown transform: {transform!r}")
            value = fn(value, arg)
            if trace_out is not None:
                pipe_steps.append({
                    "transform": transform,
                    "arg": arg,
                    "input": step_input,
                    "output": value,
                })

        final = self._truncate(value)
        if trace_out is not None:
            trace_out.append({
                "kind": "placeholder",
                "raw": raw,
                "group_name": name,
                "initial_value": initial,
                "pipes": pipe_steps,
                "final_value": final,
            })
        return final

    # ------------------------------------------------------------------
    # Utilities
    # ------------------------------------------------------------------
    @classmethod
    def _truncate(cls, value: str) -> str:
        if len(value) > cls.MAX_INPUT_LEN:
            return value[: cls.MAX_INPUT_LEN]
        return value


# ---------------------------------------------------------------------------
# Convenience wrapper matching the API the tests expect.
# ---------------------------------------------------------------------------
def render(template: str, groups: dict[str, Any]) -> str:
    return TemplateEngine().render(template, groups)

File: frontend/src/utils/templateEngine.ts
/**
 * Template engine for dummy EPG title/description rendering.
 *
 * Mirrors backend/template_engine.py — any syntax change must be made in
 * both to keep live-preview output in sync with server-rendered output.
 *
 * Syntax:
 *   - Placeholders          {name}
 *   - Chained pipes         {name|uppercase|trim|strip:-}
 *   - Conditionals          {if:group}body{/if}
 *                           {if:group=value}body{/if}
 *                           {if:group~regex}body{/if}
 *   - Legacy suffix         {name_normalize}  (lowercase, strip non-alphanumeric)
 */

export class TemplateSyntaxError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TemplateSyntaxError';
  }
}

// ---------------------------------------------------------------------------
// Caps (mirror of the Python values)
// ---------------------------------------------------------------------------
export const MAX_TEMPLATE_LEN = 4096;
export const MAX_INPUT_LEN = 1024;
export const MAX_REGEX_LEN = 500;

// ---------------------------------------------------------------------------
// Legacy normalize — lowercase + strip everything that isn't a-z or 0-9.
// ---------------------------------------------------------------------------
function legacyNormalize(value: string): string {
  return value.toLowerCase().replace(/[^a-z0-9]/g, '');
}

// ---------------------------------------------------------------------------
// Transforms
// ---------------------------------------------------------------------------
type Transform = (value: string, arg: string | null) => string;

const TRANSFORMS: Record<string, Transform> = {
  uppercase: (v) => v.toUpperCase(),
  lowercase: (v) => v.toLowerCase(),
  titlecase: (v) => v.replace(/\b\w/g, (c) => c.toUpperCase()).replace(/\B\w/g, (c) => c.toLowerCase()),
  trim: (v) => v.trim(),
  strip: (v, arg) => stripChars(v, arg),
  replace: (v, arg) => replaceArg(v, arg),
  normalize: (v) => legacyNormalize(v),
};

function stripChars(value: string, arg: string | null): string {
  if (!arg) return value.trim();
  // Build a class of chars to strip from both ends.
  const escaped = arg.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const re = new RegExp(`^[${escaped}]+|[${escaped}]+$`, 'g');
  return value.replace(re, '');
}

function replaceArg(value: string, arg: string | null): string {
  if (arg === null) return value;
  let src: string;
  let dst: string;
  const idx = arg.indexOf(':');
  if (idx === -1) {
    src = arg;
    dst = '';
  } else {
    src = arg.slice(0, idx);
    dst = arg.slice(idx + 1);
  }
  // Literal (non-regex) global replace.
  return value.split(src).join(dst);
}

// ---------------------------------------------------------------------------
// TemplateEngine
// ---------------------------------------------------------------------------
/** Shape of a single entry in the trace returned by `renderWithTrace`. */
export type TraceStep =
  | { kind: 'literal'; text: string }
  | {
      kind: 'placeholder';
      raw: string;
      group_name: string;
      initial_value: string;
      pipes: PipeStep[];
      final_value: string;
    }
  | {
      kind: 'conditional';
      condition: string;
      kind_detail: 'truthy' | 'equality' | 'regex';
      taken: boolean;
      value: string;
      body: TraceStep[];
    };

export interface PipeStep {
  transform: string;
  arg: string | null;
  input: string;
  output: string;
  /** Provenance note for synthesised steps (e.g. the legacy _normalize suffix). */
  source?: string;
}

export class TemplateEngine {
  render(
    template: string,
    groups: Record<string, unknown>,
  ): string {
    if (template === null || template === undefined) return '';
    if (template.length > MAX_TEMPLATE_LEN) {
      throw new TemplateSyntaxError(
        `Template exceeds maximum length of ${MAX_TEMPLATE_LEN} chars`,
      );
    }

    const boundedGroups: Record<string, string> = {};
    for (const [k, v] of Object.entries(groups)) {
      boundedGroups[k] = truncate(String(v ?? ''));
    }
    return this.renderSegment(template, boundedGroups);
  }

  renderWithTrace(
    template: string,
    groups: Record<string, unknown>,
  ): { output: string; trace: TraceStep[] } {
    if (template === null || template === undefined) return { output: '', trace: [] };
    if (template.length > MAX_TEMPLATE_LEN) {
      throw new TemplateSyntaxError(
        `Template exceeds maximum length of ${MAX_TEMPLATE_LEN} chars`,
      );
    }
    const boundedGroups: Record<string, string> = {};
    for (const [k, v] of Object.entries(groups)) {
      boundedGroups[k] = truncate(String(v ?? ''));
    }
    const trace: TraceStep[] = [];
    const output = this.renderSegment(template, boundedGroups, trace);
    return { output, trace };
  }

  private renderSegment(
    template: string,
    groups: Record<string, string>,
    traceOut?: TraceStep[],
  ): string {
    const out: string[] = [];
    let i = 0;
    const n = template.length;

    while (i < n) {
      const brace = template.indexOf('{', i);
      if (brace === -1) {
        const tail = template.slice(i);
        out.push(tail);
        if (traceOut && tail) traceOut.push({ kind: 'literal', text: tail });
        break;
      }

      if (brace > i) {
        const literal = template.slice(i, brace);
        out.push(literal);
        if (traceOut) traceOut.push({ kind: 'literal', text: literal });
      }

      const close = template.indexOf('}', brace);
      if (close === -1) {
        throw new TemplateSyntaxError(`Unclosed '{' at position ${brace}`);
      }

      const directive = template.slice(brace + 1, close);

      if (directive.startsWith('if:')) {
        const bodyStart = close + 1;
        const [bodyEnd, after] = this.findMatchingEndif(template, bodyStart);
        const condition = directive.slice(3);
        const { taken, detail } = this.evaluateCondition(condition, groups);
        const bodyTrace: TraceStep[] = [];
        if (taken) {
          out.push(this.renderSegment(
            template.slice(bodyStart, bodyEnd),
            groups,
            traceOut ? bodyTrace : undefined,
          ));
        }
        if (traceOut) {
          traceOut.push({
            kind: 'conditional',
            condition,
            kind_detail: detail.kind,
            taken,
            value: detail.value,
            body: bodyTrace,
          });
        }
        i = after;
        continue;
      }

      if (directive === '/if') {
        throw new TemplateSyntaxError("Unmatched '{/if}'");
      }

      out.push(this.renderPlaceholder(directive, groups, traceOut));
      i = close + 1;
    }

    return out.join('');
  }

  private findMatchingEndif(template: string, bodyStart: number): [number, number] {
    let depth = 1;
    let i = bodyStart;
    const n = template.length;
    while (i < n) {
      const b = template.indexOf('{', i);
      if (b === -1) break;
      const c = template.indexOf('}', b);
      if (c === -1) break;
      const directive = template.slice(b + 1, c);
      if (directive.startsWith('if:')) depth += 1;
      else if (directive === '/if') {
        depth -= 1;
        if (depth === 0) return [b, c + 1];
      }
      i = c + 1;
    }
    throw new TemplateSyntaxError("'{if:...}' without matching '{/if}'");
  }

  private evaluateCondition(
    condition: string,
    groups: Record<string, string>,
  ): { taken: boolean; detail: { kind: 'truthy' | 'equality' | 'regex'; value: string } } {
    // Equality check first — but only if the '=' comes before any '~'.
    const eqIdx = condition.indexOf('=');
    const tildeIdx = condition.indexOf('~');
    if (eqIdx !== -1 && (tildeIdx === -1 || eqIdx < tildeIdx)) {
      const name = condition.slice(0, eqIdx);
      const expected = condition.slice(eqIdx + 1);
      const value = groups[name] ?? '';
      return { taken: value === expected, detail: { kind: 'equality', value } };
    }

    if (tildeIdx !== -1) {
      const name = condition.slice(0, tildeIdx);
      const pattern = condition.slice(tildeIdx + 1);
      const value = groups[name] ?? '';
      if (pattern.length > MAX_REGEX_LEN) {
        return { taken: false, detail: { kind: 'regex', value } };
      }
      try {
        const re = new RegExp(pattern);
        return { taken: re.test(value), detail: { kind: 'regex', value } };
      } catch {
        return { taken: false, detail: { kind: 'regex', value } };
      }
    }

    const value = groups[condition] ?? '';
    return { taken: Boolean(value), detail: { kind: 'truthy', value } };
  }

  private renderPlaceholder(
    body: string,
    groups: Record<string, string>,
    traceOut?: TraceStep[],
  ): string {
    const parts = body.split('|');
    const name = parts[0].trim();
    const pipes = parts.slice(1);
    const raw = '{' + body + '}';

    if (name.endsWith('_normalize') && !(name in groups)) {
      const base = name.slice(0, -'_normalize'.length);
      const initial = groups[base] ?? '';
      const final = legacyNormalize(initial);
      if (traceOut) {
        traceOut.push({
          kind: 'placeholder',
          raw,
          group_name: base,
          initial_value: initial,
          pipes: [{
            transform: 'normalize',
            arg: null,
            input: initial,
            output: final,
            source: 'legacy _normalize suffix',
          }],
          final_value: final,
        });
      }
      return final;
    }

    const initial = groups[name] ?? '';
    let value = initial;
    const pipeSteps: PipeStep[] = [];

    for (const rawPipe of pipes) {
      const pipeSpec = rawPipe.trim();
      if (!pipeSpec) continue;

      const colonIdx = pipeSpec.indexOf(':');
      const transform = colonIdx === -1 ? pipeSpec : pipeSpec.slice(0, colonIdx);
      const arg = colonIdx === -1 ? null : pipeSpec.slice(colonIdx + 1);
      const stepInput = value;

      const fn = TRANSFORMS[transform];
      if (!fn) throw new TemplateSyntaxError(`Unknown transform: "${transform}"`);
      value = fn(value, arg);
      if (traceOut) {
        pipeSteps.push({ transform, arg, input: stepInput, output: value });
      }
    }

    const final = truncate(value);
    if (traceOut) {
      traceOut.push({
        kind: 'placeholder',
        raw,
        group_name: name,
        initial_value: initial,
        pipes: pipeSteps,
        final_value: final,
      });
    }
    return final;
  }
}

function truncate(value: string): string {
  return value.length > MAX_INPUT_LEN ? value.slice(0, MAX_INPUT_LEN) : value;
}

export function render(
  template: string,
  groups: Record<string, unknown>,
): string {
  return new TemplateEngine().render(template, groups);
}
